crates/egui/src/context.rs

```diff
@@ -146,24 +154,35 @@ struct ContextImpl {
 
     os: OperatingSystem,
 
-    input: InputState,
+    input: ViewportMap<InputState>,
 
     /// State that is collected during a frame and then cleared
-    frame_state: FrameState,
+    frame_state: ViewportMap<FrameState>,
+
+    /// How deeply nested are we?
+    viewport_stack: Vec<ViewportIdPair>,
 
     // The output of a frame:
-    graphics: GraphicLayers,
-    output: PlatformOutput,
+    graphics: ViewportMap<GraphicLayers>,
+    output: ViewportMap<PlatformOutput>,
 
     paint_stats: PaintStats,
 
     repaint: Repaint,
 
+    viewports: ViewportMap<Viewport>,
+    viewport_commands: Vec<(ViewportId, ViewportCommand)>,
+
+    is_desktop: bool,
+    force_embedding: bool,
+
     /// Written to during the frame.
     layer_rects_this_frame: ahash::HashMap<LayerId, Vec<(Id, Rect)>>,
+    layer_rects_this_viewports: ViewportMap<HashMap<LayerId, Vec<(Id, Rect)>>>,
```

Comment:
Why do we need to duplicate these and others? Isn't it enough to use `layer_rects_this_viewports[current_viewport_id]` everywhere? It would mean less bookkeeping at the start and end of each frame

Reply:
We can!
Was like that from the stage when i was trying to make stuff to work!

I will do that now!